### PR TITLE
feat(bb.js): initial API

### PR DESCRIPTION
**Pull request by ludamad**
# Description

Barretenberg.js initial bindings. Mainly exposes pedersen. To make this work you should also merge `adam/feat/typescript-bindings-yalc` into the repo. It is not included here as it is many files that would distract the review. They should probably be merged together.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -3,6 +3,129 @@
...
+    "__bits": "cpp",
+    "charconv": "cpp",
+    "span": "cpp",
+    "*.inc": "cpp",
+    "*.macros": "cpp"
+  },
+  "solidity.compileUsingRemoteVersion": "v0.6.10+commit.00c0fcaf",
+  "search.exclude": {
+    "**/.yarn": true,
+    "**/.pnp.*": true
```
Comment (ludamad) [inline .vscode/settings.json:112]: Settings for tsdk (at least if we decided to keep using PNP), prettier, eslint

File: cpp/src/CMakeLists.txt [hunk 2]
```diff
@@ -79,11 +80,16 @@ if(WASM)
...
         $<TARGET_OBJECTS:polynomials_objects>
         $<TARGET_OBJECTS:plonk_objects>
+        $<TARGET_OBJECTS:honk_objects>
+        $<TARGET_OBJECTS:proof_system_objects>
         $<TARGET_OBJECTS:stdlib_primitives_objects>
         $<TARGET_OBJECTS:stdlib_schnorr_objects>
         $<TARGET_OBJECTS:stdlib_pedersen_objects>
         $<TARGET_OBJECTS:stdlib_blake2s_objects>
         $<TARGET_OBJECTS:stdlib_blake3s_objects>
+        $<TARGET_OBJECTS:stdlib_sha256_objects>
```
Comment (ludamad) [inline cpp/src/CMakeLists.txt:90]: Missing from wasm build

File: ts/.pnp.cjs [hunk 3]
```diff
@@ -0,0 +1,17405 @@
+#!/usr/bin/env node
+/* eslint-disable */
+"use strict";
+
```
Comment (ludamad) [inline ts/.pnp.cjs:4]: PNP boilerplate

File: ts/src/crypto/pedersen/pedersen.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ */
+export function pedersenCompressWithHashIndex(
+  wasm: BarretenbergWasm,
+  inputs: Buffer[],
+  hashIndex: number
+): Buffer {
+  // If not done already, precompute constants.
+  wasm.call("pedersen__init");
+  const inputVectors = serializeBufferArrayToVector(inputs);
+  wasm.writeMemory(inputVectors, 0);
```
Comment (ludamad) [inline ts/src/crypto/pedersen/pedersen.ts:64]: Hm. Should we be writing to 0 instead of using bbmalloc?

File: ts/src/crypto/pedersen/pedersen.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import { Buffer } from "buffer";
+
+/**
+ * Combines two 32-byte hashes.
+ * @param wasm - The barretenberg module.
+ * @param lhs - The first hash.
+ * @param rhs - The second hash.
+ * @returns The new 32-byte hash.
+ */
+export function pedersenCompress(
```
Comment (sirasistant) [inline ts/src/crypto/pedersen/pedersen.ts:16]: I like the function format a lot for these wasm calls

File: ts/src/crypto/pedersen/pedersen.ts (hunk 4 above)
Comment (LeilaWang) [reply ts/src/crypto/pedersen/pedersen.ts:64]: For data we can't control how big it will be should be written to allocated memory?

Comment (ludamad) [reply ts/src/crypto/pedersen/pedersen.ts:64]: My thought too. I did a straight refactor for now, but some stuff has overflow risks

**Pull request merged**